cpp/src/arrow/engine/expression.h

```diff
@@ -0,0 +1,490 @@
+    if (left == NULLPTR || right == NULLPTR) {
+      return Status::Invalid("Compare operands must be non-nulls");
+    }
+
+    // Broadcast ensures that types are compatible in shape and type.
+    auto broadcast = ExprType::Broadcast(left->type(), right->type());
+    // The type of comparison is always a boolean predicate.
+    auto cast = [](const ExprType& t) { return t.CastTo(boolean()); };
+    ARROW_ASSIGN_OR_RAISE(auto type, broadcast.Map(cast));
+
+    return std::shared_ptr<Self>(
+        new Self(std::move(type), std::move(left), std::move(right)));
+  }
+
+ protected:
+  CmpOpExpr(ExprType type, std::shared_ptr<Expr> left, std::shared_ptr<Expr> right)
+      : BinaryOpExpr(std::move(left), std::move(right)),
+        Expr(expr_traits<Self>::kind_id, std::move(type)) {}
+};
+
+class ARROW_EXPORT EqualCmpExpr : public CmpOpExpr<EqualCmpExpr> {
+ protected:
+  using CmpOpExpr<EqualCmpExpr>::CmpOpExpr;
+};
+
+class ARROW_EXPORT NotEqualCmpExpr : public CmpOpExpr<NotEqualCmpExpr> {
+ protected:
+  using CmpOpExpr<NotEqualCmpExpr>::CmpOpExpr;
+};
+
+class ARROW_EXPORT GreaterThanCmpExpr : public CmpOpExpr<GreaterThanCmpExpr> {
+ protected:
+  using CmpOpExpr<GreaterThanCmpExpr>::CmpOpExpr;
+};
+
+class ARROW_EXPORT GreaterEqualThanCmpExpr : public CmpOpExpr<GreaterEqualThanCmpExpr> {
+ protected:
+  using CmpOpExpr<GreaterEqualThanCmpExpr>::CmpOpExpr;
+};
+
+class ARROW_EXPORT LessThanCmpExpr : public CmpOpExpr<LessThanCmpExpr> {
+ protected:
+  using CmpOpExpr<LessThanCmpExpr>::CmpOpExpr;
+};
+
+class ARROW_EXPORT LessEqualThanCmpExpr : public CmpOpExpr<LessEqualThanCmpExpr> {
+ protected:
+  using CmpOpExpr<LessEqualThanCmpExpr>::CmpOpExpr;
+};
+
+//
+// Relational Expressions
+//
+
+template <typename Self>
+class ARROW_EXPORT RelExpr : public Expr {
+ public:
+  const std::shared_ptr<Schema>& schema() const { return schema_; }
+
+ protected:
+  explicit RelExpr(std::shared_ptr<Schema> schema)
+      : Expr(expr_traits<Self>::kind_id, ExprType::Table(schema)),
+        schema_(std::move(schema)) {}
+
+  std::shared_ptr<Schema> schema_;
+};
+
+class ARROW_EXPORT EmptyRelExpr : public RelExpr<EmptyRelExpr> {
+ public:
+  static Result<std::shared_ptr<EmptyRelExpr>> Make(std::shared_ptr<Schema> schema);
+
+ protected:
+  using RelExpr<EmptyRelExpr>::RelExpr;
+};
+
+class ARROW_EXPORT ScanRelExpr : public RelExpr<ScanRelExpr> {
+ public:
+  static Result<std::shared_ptr<ScanRelExpr>> Make(Catalog::Entry input);
+
+  const Catalog::Entry& input() const { return input_; }
+
+ private:
+  explicit ScanRelExpr(Catalog::Entry input);
+
+  Catalog::Entry input_;
+};
+
+class ARROW_EXPORT ProjectionRelExpr : public UnaryOpExpr,
+                                       public RelExpr<ProjectionRelExpr> {
+ public:
+  static Result<std::shared_ptr<ProjectionRelExpr>> Make(
+      std::shared_ptr<Expr> input, std::vector<std::shared_ptr<Expr>> expressions);
+
+  const std::vector<std::shared_ptr<Expr>> expressions() const { return expressions_; }
```

Comment:
Preface: I am assuming that ProjectionRelExpr will always have shape TABLE, is that incorrect?

I would have expected ProjectionRelExpr to be parametrized not by a `vector<Expr>` but by `ordered_map<string, Expr>` or `ordered_map<FieldRefExpr, Expr>`. (Maybe you plan to have the key value pairs be implicit in an equality expression?)

For example
```js
project(some_table, {
  a: field("alpha"), // result's column "a" will be column "alpha" from some_table
  b: scalar(3), // result's column "b" will be the scalar 3 broadcast with "a"
})
```

Reply:
No, I'll have an `AliasExpr` for this.

```
sqlite> SELECT * FROM tmp;                                                                                                                                                                                                                                                                
a|b                                                                                                                                          
1|2                                                                                                                                                                                                                                                                  
sqlite> SELECT a+1, b FROM tmp;                                                                                                                                                                                                                                                           
a+1|b                                                                                                                                                                                                                                                                                     
2|2                                                                                                                                                                                                                                                                                       
sqlite> SELECT a+1, a+b FROM tmp;                                                                                                                                                                                                                                                         
a+1|a+b                                                                                                                                                                                                                                                                                   
2|3                                                                                                                                                                                                                                                                                                                                                                                                   
sqlite> SELECT AVG(a) + 1, a+b FROM tmp;                                                                                                                                                                                                                                                  
AVG(a) + 1|a+b                                                                                                                               
2.0|3                                                                                                                                        
sqlite> SELECT AVG(a) + 1 as mean_plus_one, a+b FROM tmp;                                                                                                                                                                                                                                 
mean_plus_one|a+b                                                                                                                            
2.0|3
```

Reply:
And to answer your question, all RelExpr will have Table shape (expect maybe Aggregate which can reduce to a single scalar), that's their common point.

Reply:
Okay, please add AliasExpr (or a TODO in the doccomment here). Out of curiosity, what will the field name for a projected Expr be if it is neither an AliasExpr nor a FieldRefExpr?